### PR TITLE
[autodiff] Support for-loop and mutation for forward mode

### PR DESCRIPTION
**Pull request by erizmr**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #5231
* __->__ #5225
* #5224

### CODE REVIEW

File: taichi/transforms/auto_diff.cpp [hunk 1]
```diff
@@ -1143,6 +1143,43 @@ class MakeDual : public ADTransform {
...
+    accumulate(stmt, dual(stmt->src.data[0].var));
+  }
+
+  void visit(LocalStoreStmt *stmt) override {
+    // Clear the dual of the dest before local store,
+    // Because LocalStoreStmt overwrites the dest,
+    // If the alloca serves as the dest of multiple LocalStoreStmt, only the
+    // last LocalStoreStmt should be taken account of, i.e, its history should
+    // be cleared
+    if (needs_grad(stmt->dest->ret_type)) {
```
Comment (ailzhang) [inline taichi/transforms/auto_diff.cpp:1174]: do we still need unified `needs_grad` interface?

File: tests/python/test_ad_for_fwd.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            p[i] = ret
+
+    for i in range(N):
+        a[i] = 3
+        b[i] = i
+
+    compute_sum()
+
+    for i in range(N):
+        assert p[i] == 3 * b[i] + 1
```
Comment (ailzhang) [inline tests/python/test_ad_for_fwd.py:28]: a[i] * b[i] +1?

File: tests/python/test_ad_for_fwd.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    with ti.ad.FwdMode(loss=p, parameters=a, seed=[1.0 for _ in range(N)]):
+        compute_sum()
+
+    for i in range(N):
+        assert p.dual[i] == b[i]
+
```
Comment (ailzhang) [inline tests/python/test_ad_for_fwd.py:35]: shall we add another FwdMode to test agaist clear_dual and non-all-one seeds? ;)

**Review by ailzhang — APPROVED** [review]
LGTM!

File: taichi/transforms/auto_diff.cpp (hunk 1 above)
Comment (erizmr) [reply taichi/transforms/auto_diff.cpp:1174]: I think this `needs_grad` is used to prevent creating `alloca` of `alloca`?

Comment (ailzhang) [reply taichi/transforms/auto_diff.cpp:1174]: Ah I think I misunderstood the `needs_grad` here (I was thinking maybe it's more proper to have `needs_dual`, then realized it only means `is_real(dt)`....

Comment (erizmr) [reply taichi/transforms/auto_diff.cpp:1174]: It’s indeed confusing here. IOI

**Pull request merged**